app/src/main/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmReceiver.kt

```diff
@@ -0,0 +1,211 @@
+            putExtra(BundleKeys.ALARM_SESSION_ID, sessionId)
+            putExtra(BundleKeys.ALARM_DAY, day)
+            putExtra(BundleKeys.ALARM_TITLE, title)
+            putExtra(BundleKeys.ALARM_START_TIME, startTime)
+            action = intentAction
+            data = Uri.parse("alarm://$sessionId")
+        }
+    }
+
+    private val logging = Logging.get()
+
+    override fun onReceive(context: Context, intent: Intent) {
+        logging.d(LOG_TAG, "Received alarm = " + intent.action + ".")
+
+        when (intent.action){
+            ALARM_DISMISSED -> onSessionAlarmNotificationDismissed(intent)
+            ALARM_UPDATE -> start(context)
+            ALARM_SESSION -> {
+                val sessionId = intent.getStringExtra(BundleKeys.ALARM_SESSION_ID)!!
+                val day = intent.getIntExtra(BundleKeys.ALARM_DAY, 1)
+                val start = intent.getLongExtra(BundleKeys.ALARM_START_TIME, System.currentTimeMillis())
+                val title = intent.getStringExtra(BundleKeys.ALARM_TITLE)!!
+                logging.report(LOG_TAG, "sessionId = $sessionId, intent = $intent")
+                //Toast.makeText(context, "Alarm worked.", Toast.LENGTH_LONG).show();
+
+                val uniqueNotificationId = AppRepository.createSessionAlarmNotificationId(sessionId)
+                val notificationHelper = NotificationHelper(context)
+
+                notificationHelper.notify(
+                    uniqueNotificationId,
+                    notificationHelper.getSessionAlarmNotificationBuilder(
+                        PendingIntent.getActivity(
+                            context,
+                            DEFAULT_REQUEST_CODE,
+                            createLaunchIntent(context, sessionId, day, uniqueNotificationId),
+                            PendingIntent.FLAG_ONE_SHOT or FLAG_IMMUTABLE
+                        ),
+                        title,
+                        start,
+                        AppRepository.readAlarmToneUri(),
+                        PendingIntent.getBroadcast(
+                            context,
+                            DEFAULT_REQUEST_CODE,
+                            createDeleteNotificationIntent(context, uniqueNotificationId),
+                            FLAG_IMMUTABLE
+                        )
+                    ),
+                    AppRepository.readInsistentAlarmsEnabled()
+                )
+                AppRepository.deleteAlarmForSessionId(sessionId)
+            }
+        }
+    }
+
+    private fun onSessionAlarmNotificationDismissed(intent: Intent) {
+        val notificationId = intent.getIntExtra(BUNDLE_KEY_NOTIFICATION_ID, INVALID_NOTIFICATION_ID)
+
+        check(notificationId != INVALID_NOTIFICATION_ID) {
+            "Bundle does not contain NOTIFICATION_ID."
+        }
+
+        AppRepository.deleteSessionAlarmNotificationId(notificationId)
+    }
+
+    class AlarmIntentBuilder {
```

Comment:
This builder is actually not a builder but a factory. All parameter are mandatory anyway. Let's rewrite this:

``` kotlin
internal class AlarmIntentFactory(
    val context: Context,
    val sessionId: String,
    val title: String,
    val day: Int,
    val startTime: Long,
) {

    fun getIntent(isAddAlarmIntent: Boolean) = if (isAddAlarmIntent) {
        getAlarmIntent(context, sessionId, day, title, startTime, ...)
    } else {
        getAlarmIntent(context, sessionId, day, title, startTime, ...)
    }

    // ... move helper function here
    // ... move constants which are only used in the class here

}
```

Usage:

``` kotlin
val intent = AlarmReceiver.AlarmIntentFactory(
    context = context,
    sessionId = alarm.sessionId,
    day = alarm.day,
    title = alarm.sessionTitle,
    startTime = alarm.startTime
).getIntent(isAddAlarmIntent = true)
```

Much more concise :)